### PR TITLE
Add Lion optimizer

### PR DESCRIPTION
**Pull request by drauh**
Add Lion optimizer to keras, using the experimental optimizer API. 
Authors [implementation](http://github.com/google/automl/tree/master/lion) is based on legacy optimizer API.

Request for contribution was made [here](https://github.com/keras-team/keras/issues/17570)

### CODE REVIEW

**Comment by innat** [comment]
cc. @crazydonkey200

File: keras/optimizers/lion.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+from keras.optimizers import optimizer
+from keras.saving.object_registration import register_keras_serializable
+
+# isort: off
+from tensorflow.python.util.tf_export import keras_export
+
+
+@register_keras_serializable()
+@keras_export(
+    "keras.optimizers.experimental.Lion",
```
Comment (fchollet) [inline keras/optimizers/lion.py:28]: This is not needed

File: keras/optimizers/lion.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        that takes no arguments and returns the actual value to use. The rate
+        to combine the current gradient and the 1st moment estimate.
+      beta_2: A float value or a constant float tensor, or a callable
+        that takes no arguments and returns the actual value to use. The
+        exponential decay rate for the 1st moment estimate.
+    {{base_optimizer_keyword_args}}
+
+    References:
+      - [Chen et al., 2023](http://arxiv.org/abs/2302.06675)
+      - [Authors implementation](\
```
Comment (fchollet) [inline keras/optimizers/lion.py:57]: The backslash is not needed

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

CC @chenmoneygithub 

Should we add an optimizer-specific correctness test?

File: keras/optimizers/lion.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import tensorflow.compat.v2 as tf
+
+from keras.optimizers import optimizer
+from keras.saving.object_registration import register_keras_serializable
+
+# isort: off
+from tensorflow.python.util.tf_export import keras_export
+
+
+@register_keras_serializable()
```
[the remark this answers is not in the document]
Comment (drauh) [reply keras/optimizers/lion.py]: removed

File: keras/optimizers/lion.py (hunk 1 above)
Comment (drauh) [reply keras/optimizers/lion.py]: removed

File: keras/optimizers/lion.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+@keras_export(
+    "keras.optimizers.experimental.Lion",
+    "keras.optimizers.Lion",
+    v1=[],
+)
+class Lion(optimizer.Optimizer):
+    r"""Optimizer that implements the Lion algorithm.
```
[the remark this answers is not in the document]
Comment (drauh) [reply keras/optimizers/lion.py]: removed

File: keras/optimizers/lion.py (hunk 2 above)
Comment (drauh) [reply keras/optimizers/lion.py]: removed

**Comment by chenmoneygithub** [comment]
@fchollet Do we want to verify the numerical correctness or convergence? The previous one is a little hard, as the golden value should come from manual calculation.

**Review by chenmoneygithub — COMMENTED** [review]
Thanks for the PR, it looks pretty nice! 

By the way, since you are the first OSS contributor who writes a new optimizer based on our new API, may I ask how do you feel about our new API? Is it easy to follow and customize?

**Comment by drauh** [comment]
Thanks ! Yes It was easy, I used other implementations (SGD, Adamax) to help. To only have 3 methods to implement instead of 5 is nice. Especially not needing the `_prepare_local` method, to use `get_slot` (even if `_index_dict` is still a bit mysterious to me), or to `control_dependencies` simplifies a lot. Having decoupled weight decay, ema and gradient cliping instantly is a nice bonus.

Regarding ema, i'm still using custom code because i like to swap weights for evaluation and at the end of the training, and not overwriting periodically, as from what I know, it's not the standard way to do polyak averaging (but it could be better, i don't know!).

**Comment by fchollet** [comment]
> Regarding ema, i'm still using custom code because i like to swap weights for evaluation and at the end of the training, and not overwriting periodically, as from what I know, it's not the standard way to do polyak averaging (but it could be better, i don't know!).

You can still do this with the new API. The optimizer lets you merge the weights manually the end of training.

**Comment by drauh** [comment]
Yes sure it does it at the end of the training. But a method `swap_weights` like the one of `MovingAverage` optimizer of tensorflow_addons, coupled with a callback to swap before and swap back after validation allows to calculate val losses and metrics on the average weights at every epochs without any overwrite.

**Comment by crazydonkey200** [comment]
Thanks for adding the lion optimizer! The implementation looks great. 

One question: In our original implementation, we had decoupled weight decay similar to AdamW in the optimizer. Do we need to add that here or is the best practice to handle it outside the optimizer?

**Comment by crazydonkey200** [comment]
> @fchollet Do we want to verify the numerical correctness or convergence? The previous one is a little hard, as the golden value should come from manual calculation.

One thing we did before to test the numerical correctness is to compare the results with a simple numpy implementation after running the optimizer for a few steps, similar to [this](https://github.com/tensorflow/tensorflow/blob/v2.11.0/tensorflow/python/training/adam_test.py#L35).

**Comment by drauh** [comment]
> Thanks for adding the lion optimizer! The implementation looks great.
> 
> One question: In our original implementation, we had decoupled weight decay similar to AdamW in the optimizer. Do we need to add that here or is the best practice to handle it outside the optimizer?

@crazydonkey200 Thanks ! 

All keras optimizers (new API) implement a decoupled weight decay, see [here](https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/optimizers/optimizer_experimental/optimizer.py#L1152). It should yield exactly the same result as your implementation.

Also I changed the implementation of the sparse update from what you did [here](https://github.com/google/automl/blob/master/lion/lion_tf2.py) using that :
image
Thus, we can combine momentums and gradients with beta_1 and use this to update model parameters and then combine momentums and gradients with beta_2 / beta_1 to update momentums. It should be mathematically equal but removes one update.

**Comment by xiangning-chen** [comment]
@drauh Thanks for the implementation!

One comment is that do you need to add the weight_decay in the config here: https://github.com/keras-team/keras/pull/17605/files#diff-be32717e922f2f1ee840edf1266148a3a103ed40a31cadd59429d722f0193660R139

Cause it's included in the AdamW implementation: https://github.com/keras-team/keras/blob/master/keras/optimizers/adamw.py#L216

**Comment by drauh** [comment]
It is already added in the [parent class](https://github.com/keras-team/keras/blob/0e0d77fd5f1544545980e9388bb0264751e01997/keras/optimizers/optimizer.py#L757).

In fact `AdamW = Adam` only with a default weight decay of 0.0004 and a different docstring. According to [this](https://github.com/keras-team/keras/commit/38b618ad90d669c85cccee521ad73cc0630cf750) commit :
> We still keep adamw optimizer in case people want an explicit adamw. We can delete it in a followup cl.

You can also see that by the duplicated weight decay parameter in the AdamW [docstring](https://keras.io/api/optimizers/adamw/), one coming from the parent class optimizer and one from the child class.
image

**Comment by xiangning-chen** [comment]
Thanks for the explanation!

Another thing is that can you please add the instructions for the recommended learning rate and weight decay value, given that the scale differs a lot from other optimizers.
Please see the optax example here: https://github.com/deepmind/optax/blob/master/optax/_src/alias.py#L319. Thank you so much!

**Comment by fchollet** [comment]
> One thing we did before to test the numerical correctness is to compare the results with a simple numpy implementation after running the optimizer for a few steps, similar to [this](https://github.com/tensorflow/tensorflow/blob/v2.11.0/tensorflow/python/training/adam_test.py#L35).

Can we add a similar correctness test in this PR?

**Comment by drauh** [comment]
Yes I can do it.

**Comment by drauh** [comment]
@fchollet @chenmoneygithub I tried to implement `beta_1 = 0` as requested.
> One minor detail is to maybe add some safeguard for the rare case where beta1=0.

However there is something not quite clear to me. Will Lion also be affected by this [bug](https://github.com/keras-team/keras/commit/91fcaf6f5fefa396b298e8131d207082061dde9d). It seem that in the SGD implementation, momentums are always [created](https://github.com/keras-team/keras/blob/5ce6017623a61e7f34daa11569e462ef8fc3f660/keras/optimizers/sgd.py#L144) and [updated](https://github.com/keras-team/keras/blob/5ce6017623a61e7f34daa11569e462ef8fc3f660/keras/optimizers/sgd.py#L167). Should we create the momentums even if they are not used ? Or they should be created but their update could be skipped.

**Comment by crazydonkey200** [comment]
> @fchollet @chenmoneygithub I tried to implement `beta_1 = 0` as requested.
> 
> > One minor detail is to maybe add some safeguard for the rare case where beta1=0.
> 
> However there is something not quite clear to me. Will Lion also be affected by this [bug](https://github.com/keras-team/keras/commit/91fcaf6f5fefa396b298e8131d207082061dde9d). It seem that in the SGD implementation, momentums are always [created](https://github.com/keras-team/keras/blob/5ce6017623a61e7f34daa11569e462ef8fc3f660/keras/optimizers/sgd.py#L144) and [updated](https://github.com/keras-team/keras/blob/5ce6017623a61e7f34daa11569e462ef8fc3f660/keras/optimizers/sgd.py#L167). Should we create the momentums even if they are not used ? Or they should be created but their update could be skipped.

@drauh  Thanks for the quick update! By safety guard, I was thinking of a simpler alternative to just assert beta1 != 0 and say that it will degenerate to SignSGD when beta1=0. Then you can keep the rest of the code intact since the current implementation is quite clean and elegant.

**Comment by drauh** [comment]
I've added the numerical correctness test and reverted the changes that allowed beta_1 = 0, replaced by an assert.

**Comment by drauh** [comment]
> This is not needed

@fchollet since I removed the decorators, there is one test that is not passing:

```
ERROR: testSaveAndLoadOptimizerWithModel7 (optimizer_fn=lion) (__main__.OptimizerFuntionalityTest)
OptimizerFuntionalityTest.testSaveAndLoadOptimizerWithModel7 (optimizer_fn=lion)
testSaveAndLoadOptimizerWithModel(optimizer_fn=lion)
ValueError: Unknown optimizer: 'Lion'. Please ensure you are using a `keras.utils.custom_object_scope` and that this object is included in the scope. See https://www.tensorflow.org/guide/keras/save_and_serialize#registering_the_custom_object for details.
```

Is that normal ?

**Comment by fchollet** [comment]
Thanks for the updates!

> @fchollet since I removed the decorators, there is one test that is not passing:

This is using the legacy SavedModel format. It appears that the new optimizer implementations aren't registered globally which is why they need the decorator. @chenmoneygithub we should register those classes globally, can you look into it?

@drauh you can revert back to using the decorator for the time being. We'll fix the issue across all new optimizers independently of this PR.

File: keras/optimizers/lion.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            jit_compile=jit_compile,
+            **kwargs,
+        )
+        self._learning_rate = self._build_learning_rate(learning_rate)
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        if self.beta_1 == 0.0:
+            raise ValueError(
+                "`beta_1` must be between [0, 1] otherwise the optimizer "
+                "degenerate to SignSGD."
```
Comment (fchollet) [inline keras/optimizers/lion.py:94]: Please print the value passed by the user.

**Comment by drauh** [comment]
Thanks for the advice, it's done !

**Review by fchollet — APPROVED** [review]
LGTM -- we should be ready to merge. Thank you for the contribution!

**Comment by xiangning-chen** [comment]
Is it possible to also add Lion to the Keras API doc? Currently I don't see it [here](https://keras.io/api/optimizers/). @fchollet

**Pull request merged**